autorest/azure/devicetoken.go

```diff
@@ -0,0 +1,196 @@
+)
+
+var (
+	// ErrDeviceGeneric represents an unknown error from the token endpoint when using device flow
+	ErrDeviceGeneric = fmt.Errorf("Error while retrieving OAuth token: Unknown Error")
+
+	// ErrDeviceAccessDenied represents an access denied error from the token endpoint when using device flow
+	ErrDeviceAccessDenied = fmt.Errorf("Error while retrieving OAuth token: Access Denied")
+
+	// ErrDeviceAuthorizationPending represents the server waiting on the user to complete the device flow
+	ErrDeviceAuthorizationPending = fmt.Errorf("Error while retrieving OAuth token: Authorization Pending")
+
+	// ErrDeviceCodeExpired represents the server timing out and expiring the code during device flow
+	ErrDeviceCodeExpired = fmt.Errorf("Error while retrieving OAuth token: Code Expired")
+
+	// ErrDeviceSlowDown represents the service telling us we're polling too often during device flow
+	ErrDeviceSlowDown = fmt.Errorf("Error while retrieving OAuth token: Slow Down")
+)
+
+// DeviceCode is the object returned by the device auth endpoint
+// It contains information to instruct the user to complete the auth flow
+type DeviceCode struct {
+	DeviceCode      *string `json:"device_code"`
+	UserCode        *string `json:"user_code"`
+	VerificationURL *string `json:"verification_url"`
+	ExpiresIn       *int64  `json:"expires_in,string"`
+	Interval        *int64  `json:"interval,string"`
+
+	Message  *string `json:"message"` // Azure specific
+	Resource *string // store this as well, needed during token exchange for azure
+}
+
+// TokenError is the object returned by the token exchange endpoint
+// when something is amiss
+type TokenError struct {
+	Error            *string `json:"error"`
+	ErrorCodes       []int   `json:"error_codes"`
+	ErrorDescription *string `json:"error_description"`
+	Timestamp        *string `json:"timestamp"`
+	TraceID          *string `json:"trace_id"`
+}
+
+// DeviceToken is the object return by the token exchange endpoint
+// It can either look like a Token or an ErrorToken, so put both here
+// and check for presence of "Error" to know if we are in error state
+type DeviceToken struct {
+	Token
+	TokenError
+}
+
+// InitiateDeviceAuth initiates a device auth flow. It returns a DeviceCode
+// that can be used with CheckForUserCompletion or WaitForUserCompletion.
+func InitiateDeviceAuth(client *autorest.Client, clientID, resource string) (*DeviceCode, error) {
+	req, _ := autorest.Prepare(
+		&http.Request{},
+		autorest.AsPost(),
+		autorest.WithBaseURL(OAuthDeviceEndpoint),
+		autorest.WithFormData(url.Values{
+			"client_id": []string{clientID},
+			"resource":  []string{resource},
+		}),
+		autorest.WithQueryParameters(map[string]interface{}{
+			authAPIVersionQueryParamName: authAPIVersionQueryParamValue,
+		}))
+
+	resp, err := client.Send(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error occurred while requesting Device Authorization Code. %s", err)
+	}
+	if resp.StatusCode != http.StatusOK {
```

Comment:
I think we can use responder decorators here to do status checks etc. (i.e. this package can have its own `withErrorUnlessStatusCode()` and you can close the body and deserialize json in that responder list easily as Brendan said.